20653: Adds parameter description

nuvious opened the pull request:
Initial draft of PR to refactor location of parameter docstrings

cademack committed:
add code to parse param descriptions

cademack committed:
Merge remote-tracking branch 'origin/main' into 20653-api-descriptions

Commit:
Starting with CONTRIBUTING.md

nuvious committed:
Incremental progress commit.

nuvious committed:
Got all the ones tagged with ;parameters:. Found others that may need adjusting that don't have a ;parameters: tag.

nuvious committed:
Found some other ones that weren't caught with my first regex.

nuvious committed:
Think I got em all.

nuvious committed:
Merge branch 'main' into 20653-AddParameterDescriptions_Cheese

cademack commented on howso/ablation.amlg:
```diff
@@ -111,23 +111,21 @@
 	(declare
 		(assoc
 			;{type "string"}
+			;optional, default !autoAblationWeightFeature. name of feature whose values to use as case weights
```
Lets remove the 'optional' from the text descriptions. In this case its actually not optional, but even in the places where the default value is (null) we will have that automatically inspected (or at least plan to).

nuvious replied:
Should be resolved, but will allow you to review and resolve comment.

cademack commented on howso/analysis.amlg:
```diff
@@ -15,57 +15,44 @@
+			;optional
 			bypass_hyperparameter_analysis (null)
 			;{type "boolean"}
+			;optional
```
feel free to leave these optionals here, I see that there was no description fo this parameter. We can use this as a pseudo signal for Res or I to come add proper descriptions

cademack replied:
"these" = parameters that didn't have a proper description for you to move into the assoc

nuvious replied:
Should I still remove the 'optional' parameters from the ones that do have full descriptions? Also I think I inadvertently dropped some default parameter statements, can't remember if you wanted those preserved or removed.

cademack replied:
Yeah let's remove 'optional' from those with full descriptions. I think we can drop the mention of defaults too, but it might make sense for some descriptions to keep them around if there is additional information to capture along the default value

cademack commented on howso/editing.amlg:
```diff
@@ -1,14 +1,112 @@
 ;Contains methods for adding, editing and removal of cases and session data.
 (null
+	;Cleans up trainee data, removing unused sessions, cases or actions missing data, etc. given the following optional parameters.  If a parameter is not specified, it will look to this entity's own label of the same name.  See label for parameter details.
+	#clean_data
```
I think this label got removed in main, might want to merge main when possible

nuvious replied:
Yup, doing this today.

nuvious replied:
Removed in next commit

cademack commented on howso/react.amlg:
```diff
@@ -30,6 +23,7 @@
 			;{type "list" values "string"}
 			post_process_features (list)
 			;{type "list" values {type "list"}}
+			;see #react for description. if specified must be either lenght of 1 or num_reacts.
```
I see it isn't your typo originally, but lets fix it now I guess

nuvious replied:
Not apposed to fixes like that and should be in next commit.

howsohazard committed:
Merge branch 'main' into 20653-AddParameterDescriptions_Cheese

Commit:
Incremental commit before doing an update from main.

Commit:
Merge branch '20653-AddParameterDescriptions_Cheese' of github.com:howsoai/howso-engine into 20653-AddParameterDescriptions_Cheese

Commit:
Merge branch 'main' into 20653-AddParameterDescriptions_Cheese

Commit:
PR Comments addressed

Commit:
Typo that broke unit tests.

nuvious commented:
Single unit test failing, but switched to main and it's failing there as well:

```
$ amalgam-st run_all_tests.amlg # Amalgam version 54.3.5
....
=====================================
datetime data_type automatically set to 'formatted_date_time': PASSED (same)
React outputs correct date: FAILED: assert_same, Expected: 2020-06-08 lunes 11.33.43  Observed: 1970-01-01 Thursday 00.00.00
Using datetime as context in react: PASSED (same), Expected: 25  Observed: 25
Number of failed asserts: 1 
[FAILED: Datetime as context.]
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

FAILED >> ut_h_datetimeformat.amlg <<
```

cademack commented on howso/get_cases.amlg:
```diff
@@ -88,39 +88,37 @@
+			;if set to 1 will return the audit data of which features for each row were imputed (auto-filled)
 			indicate_imputed (false)
 			;{type "boolean"}
+			;optional, flag. When true only returns an assoc of caseds and no features
```
I figure this should be "case ids"

nuvious replied:
Addressed in [d8a9404f922319cfd8e1c60555d514603ff45ec3](https://github.com/howsoai/howso-engine/pull/239/commits/ce2065a71458ab59dc663e8dbf54b9e9a94911dc)

cademack commented on howso/get_cases.amlg:
```diff
@@ -88,39 +88,37 @@
+			;optional string. default is 'exact', used only with 'condition' parameter, will find exact matches if 'exact' and similar cases if 'similar'.
 			precision "exact"
 			;{type "boolean"}
+			;if set to 1 will return the audit data of which features for each row were imputed (auto-filled)
```
Let's switch this to "if true, ..."

nuvious replied:
Addressed in [d8a9404f922319cfd8e1c60555d514603ff45ec3](https://github.com/howsoai/howso-engine/pull/239/commits/ce2065a71458ab59dc663e8dbf54b9e9a94911dc)

cademack commented on howso/hierarchy.amlg:
```diff
@@ -80,30 +76,28 @@
 			;{type "boolean"}
+			;optional, boolean, applicable only to 'load_subtrainee' and 'save_subtrainee' and if specifying child_name_path or child_id.
+	;	For 'save_subtrainee', stores the child out as an independent trainee and removes it as a contained entity.
+	;	For 'load_subtrainee' updates hierarchy by adding the child as an independently stored trainee to the hierarchy without loading the trainee as a subtrainee.
```
Align indentation here please

nuvious replied:
Address in [d8a9404f922319cfd8e1c60555d514603ff45ec3](https://github.com/howsoai/howso-engine/pull/239/commits/ce2065a71458ab59dc663e8dbf54b9e9a94911dc)

cademack commented on howso/update_cases.amlg:
```diff
@@ -127,42 +127,42 @@
+			; session: 
 			precision "exact"
 			;{type "string"}
+			;the session id when this call !is being made
```
dubious exclamation point

nuvious replied:
Found multiple locations with this error too, all addressed in [d8a9404f922319cfd8e1c60555d514603ff45ec3](https://github.com/howsoai/howso-engine/pull/239/commits/ce2065a71458ab59dc663e8dbf54b9e9a94911dc)

Commit:
Updated .gitignore to exclude programatically generated unit tests.

Commit:
PR comment adjustments.

Commit:
Merge branch 'main' into 20653-AddParameterDescriptions_Cheese

Commit:
Merge branch 'main' into 20653-AddParameterDescriptions_Cheese

Commit:
Merge branch 'main' into 20653-AddParameterDescriptions_Cheese

cademack commented on howso/analysis.amlg:
```diff
@@ -15,57 +15,44 @@
+			;	NOTE: if the dataset has been edited or has multiple trained sessions, the holdouts may be imbalanced due to multiple cases having
+			;	the same session id index or missing session indices, therefore this flag should only be used internally for validation
+			;{type "boolean"}
+			;optional
```
can remove the extra type {} dict and optional here

nuvious replied:
Corrected in 6d0f612

cademack commented on howso/analysis.amlg:
```diff
@@ -15,57 +15,44 @@
 			num_samples (null)
 			;{type "number"}
+			;optional. number of cases to sample during analysis. only applies for k_folds = 1
+			; analysis_sub_model_size: 
```
`; analysis_sub_model_size` shouldn't be here

nuvious replied:
Corrected in 6d0f612

cademack commented on howso/feature_conviction.amlg:
```diff
@@ -53,34 +53,32 @@
+			;list of label names
 			features (list)
 			;{type "list" values "string"}
+			;optional, list of action features to use as the baseline for conviction instead of the full model
```
I meant earlier that I don't think we need `optional` included on *any* of the parameter descriptions as we will introspect that through the default value in the code. 

However, if there is a parameter that did not have any form of docstring before, then feel free to just leave `;optional` there so Res or I can come through in a later pass and give the parameter a full description.

nuvious replied:
Sent a Teams message to ensure I understand the above and will address once I confirm I don't have a misunderstanding.

nuvious replied:
Addressed as discussed in 8f7b442

cademack commented on howso/feature_conviction.amlg:
```diff
@@ -53,34 +53,32 @@
+			;optional list of cases for which to use as the model to compute feature conviction
 			case_ids (list)
 			;{type "boolean"}
+			;deafult to true, calculate and output familiarity conviction of adding the specified features
```
typo on default here

nuvious replied:
Corrected, also found an occurence in attributes.amlg with the same typo. Corrected in 6d0f612

cademack commented on howso/react_group.amlg:
```diff
@@ -9,45 +9,43 @@
+			;default to false, calculate and output familiarity conviction of removing the specified new_cases in the output assoc
 			familiarity_conviction_removal (false)
 			;{type "boolean"}
+			;default to false, calculate and output the KL divergence of adding the specified new_cases in the output assoc
```
Where the default value is a literal/immediate value such as a string or (true)/(false), let's cut that part out of the descripition as it will be included in the API JSON automatically.

But if there is more complex behavior like "default is to use all the trained features" or something like that, please leave it. Feel free to message me about any unclear judgement calls on this

nuvious replied:
Will address with the optionals tomorrow.

nuvious replied:
Addressed in d90fe64

cademack commented on howso/remove_cases.amlg:
```diff
@@ -2,29 +2,27 @@
+			;optional, if specified, ignores condition and instead operates on all cases that were trained with this session id. Ignored if case_indices is specified.
 			condition_session (null)
 			;{ref "Precision"}
+			;flag, whether to query for 'exact' matches; if set to 'similar' will move num_cases with the most similar values. Ignored if case_indices is specified.
```
lets change 'move' to 'remove' on this line, this method was formerly "move_cases"...

nuvious replied:
Corrected in 6d0f612

cademack reviewed (commented):
We're almost there. I think we just remove the 'optional's and some 'default' mentions and we will be about done. I'm still trying to catch typos and stuff while we are doing this, but they're growing rarer - so that's good.

Commit:
Addressed PR comments.

Commit:
Removed optional qualifiers for all parameters that have a description; only 2 remain:

analysis.amlg
	bypass_calculate_feature_residuals 38
	bypass_calculate_feature_weights 41

cademack commented on howso/ablation.amlg:
```diff
@@ -111,23 +111,21 @@
 	(declare
 		(assoc
 			;{type "string"}
+			;default all trained features. a list of features to use when computing influence weight entropies
```
Looks like the wrong description got pasted here. This is the one for `features` (the below param). Should be able to grab the description for distribute_weight_feature from main still

nuvious replied:
Retreived from main and addressed in: [d54a657](https://github.com/howsoai/howso-engine/pull/239/commits/d54a6575c30a5c75415b2cc52e96d51cfb0406c6)

cademack commented on howso/feature_conviction.amlg:
```diff
@@ -53,34 +53,32 @@
 	(declare
 		(assoc
 			;{type "list" values "string"}
+			;list of label names
```
wow this is a really old description, please edit this one to "list of feature names"

nuvious replied:
Addressed in: [d54a657](https://github.com/howsoai/howso-engine/pull/239/commits/d54a6575c30a5c75415b2cc52e96d51cfb0406c6)

cademack commented on howso/get_cases.amlg:
```diff
@@ -354,7 +352,7 @@
 	;return the query conditions for cases that match the specified condition
 	;parameters:
-	; condition_session: optional, if specified ignores condition and returns cases for the specified session id
+	; condition_session: 
```
I figure this was not removed on purpose

nuvious replied:
Not intentionally for sure, restored and addressed in: [d54a657](https://github.com/howsoai/howso-engine/pull/239/commits/d54a6575c30a5c75415b2cc52e96d51cfb0406c6)

cademack commented on howso/react_aggregate.amlg:
```diff
@@ -2,153 +2,152 @@
+			;		selected_prediction_stats": list of strings, optional.  Allowed values are:
+			;						"mda" : mean decrease in accuracy when each feature is dropped from the model, applies to all features.
+			;						"contribution": feature contribution to predicted value when each feature is dropped from the model, applies to all features.
+			;						"mda_permutation": mean decrease in accuracy that used scrambling of feature values instead of dropping each feature, applies to all features.
```
These actually got removed from this list in a recent PR. Probably was hard to catch in the merge conflict, remove them here please.

nuvious replied:
Checked against main to be sure the enumerated parameters matched per your comment. Addressed in: [d54a657](https://github.com/howsoai/howso-engine/pull/239/commits/d54a6575c30a5c75415b2cc52e96d51cfb0406c6)

Commit:
Addressed PR comment to remove default statements where literal defaults are assigned such as booleans and strings.

cademack commented on howso/react_group.amlg:
```diff
@@ -9,45 +9,43 @@
 	(declare
 		(assoc
 			;{type "list" values "string"}
+			;list of label names
```
"label names" here again.. this stuff is rusty. Please change to "list of feature names"

nuvious replied:
Addressed in: [d54a657](https://github.com/howsoai/howso-engine/pull/239/commits/d54a6575c30a5c75415b2cc52e96d51cfb0406c6)

Commit:
Merge branch 'main' into 20652-AddParameterDescriptions_Cheese

Commit:
Added descriptions to optional parameters with no descriptions

Commit:
Addressed PR changes.

Commit:
Updated time series unit tests as directed by Cade.

howsoRes commented on CONTRIBUTING.md:
```diff
@@ -91,12 +91,16 @@ Return types of methods should be specified by comments on the assoc with the pa
 #foo_bar_method
 (declare
     (assoc
+        ; list, a list of strings
```
no spaces between ; and the text for any first comment, so these 4 in this file need to have that space removed

nuvious replied:
Will address pending clarification in Teams discussion.

nuvious replied:
Addressed in 5dbbf38

howsoRes commented on howso/ablation.amlg:
```diff
@@ -111,23 +111,21 @@
+			;default all trained features. a list of features to use when computing influence weight entropies
 			features !trainedFeatures
 			;{type "number"}
+			;default !autoAblationInfluenceWeightEntropyThreshold. cases with
```
remove 'cases with' at the end, it's a duped on the next line

nuvious replied:
Removed in  1632325

howsoRes commented on howso/attributes.amlg:
```diff
@@ -1,161 +1,158 @@
+			;					If there is any possibility of global relevancy of time, it is generally recommended to set this value to true, which is the default.
 			feature_attributes (assoc)
 			;{type "boolean"}
+			;flag, default to true. create time series feature attributes if necessary
```
remove 'flag,', its type is a boolean, so that's redundant. do the same for any other boolean parameter across the files.  either that, or need to add 'flag' to all booleans that don't have it specified so boolean descriptions are consistent.

nuvious replied:
Will change pending discussion in Teams

nuvious replied:
Per discussion in Teams, will be deferred for this scope of work.

howsoRes commented on howso/hyperparameters.amlg:
```diff
@@ -795,23 +790,23 @@
+			;{type "number"}
+			;the size of of the model at which to stop doing outo-analysis. Value of 0 means no limit.
+			;if parameters that are specified for analyze are passed in, they will be stored and used during auto-analysis
+			auto_analyze_limit_size 200000
```
this parameter doesn't exist so it and its comments should be removed

nuvious replied:
Removed in 1632325

nuvious committed:
Addressed PR Comments.

nuvious committed:
Bumping retry attempts to 3 for ut_h_time_series_stock based on guidance from Cade on problematic unit test.

nuvious committed:
Addressed PR comments regarding leading spaces and added clarification to contribution guide.

Commit:
Merge remote-tracking branch 'origin/main' into 20653-AddParameterDescriptions_Cheese
